### PR TITLE
Tool: Build id extractor

### PR DESCRIPTION
**Pull request by clementetb**
This tool allows building a table that binds build ids with which flavor, version, and arch.

The resulting table can help to retrieve the context information of a given native stack trace.

It accepts a single version or a version range. It works by downloading such versions from S3 and extracting the build id with the `readelf` tool from the ndk.

### CODE REVIEW

File: tools/build-id.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        f = open(path)
+        return True
+    except IOError:
+        return False
+    finally:
+        if f:
+            f.close()
+
+def find_ndk():
+    PROPS_FILE=f"{os.getcwd()}/../realm/local.properties"
```
Comment (rorbech) [inline tools/build-id.py:26]: I don't think that projects using https://developer.android.com/studio/projects/install-ndk#apply-specific-version to configure the NDK will have the `ndk.dir` set in `local.properties` so you might want to grab it by some other means. As the version is probably not that important to just read the build ids, you could maybe just grab it from the newest version available from `ANDROID_HOME` (`find $ANDROID_HOME/ndk/ -name ".*" -prune -maxdepth 1 -o -print | sort -rV | head -n 1`)

File: tools/build-id.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    major_range = parse_range(args.major)
+    minor_range = parse_range(args.minor)
+    patch_range = parse_range(args.patch)
+    
+    zips_path = get_path(os.getcwd() + "/downloads/")
+    libs_path = get_path(os.getcwd() + "/libs/")
+    build_ids = []
+
+    print("-- This process can take some minutes ⏰ --")
```
Comment (rorbech) [inline tools/build-id.py:101]: Ideally out should keep status information and program output separate. So maybe write all the debug info to stderr. This makes it way easier to use the tool to generate processable files.

File: tools/build-id.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+def find_ndk():
+    PROPS_FILE=f"{os.getcwd()}/../realm/local.properties"
+    if not file_exists(PROPS_FILE):
+        print(f"{PROPS_FILE} not found! NDK location cannot be determined")
+        exit()
+    
+    return subprocess.check_output(f'grep "ndk.dir" "{PROPS_FILE}" | cut -d = -f2', shell=True).decode('ascii').strip()
+
+
+BUILD_ID_EXTRACT_COMMMAND = find_ndk() + READELF_TOOL_PATH + ' -n {0} | grep "Build ID" | cut -d ":" -f2'
```
Comment (rorbech) [inline tools/build-id.py:34]: Maybe keep constants together at the head of the file

File: tools/build-id.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+def download_file(url, path):
+    urllib.request.urlretrieve(url, path)
+
```
Comment (rorbech) [inline tools/build-id.py:38]: Whats with all the empty lines. Is it PEP08 dictated?

**Review by rorbech — COMMENTED** [review]
Cool. I left some comments. 

Besides that, could we leverage this in our unrolling tool by identifying the build id from the stack trace and derive architecture, version, etc. from a local cache of the output from this tool.

File: tools/build-id.py (hunk 1 above)
Comment (clementetb) [reply tools/build-id.py]: Thanks, I have added this way to resolve the ndk path.

File: tools/build-id.py (hunk 4 above)
Comment (clementetb) [reply tools/build-id.py:36]: It is two empty lines spaces between functions..

**Comment by clementetb** [comment]
@rorbech We cannot directly hook it up with the unrolling tool as this script takes some time to retrieve the build ids. We could integrate the rainbow table with the unrolling tool though.

**Review by cmelchior — APPROVED** [review]
I know very litte Python, so I just assumed that if you could produce the table with it, it works as expected :D

Otherwise 🥇

**Pull request merged**